### ISSUE TITLE
[DEV] Blender json exporter, x-axis inverted

### ISSUE THREAD

**Issue by izar89** [mrdoob/three.js#10804]
blender: v2.78b
blender exporter: dev branch
three.js: r84

It seems the x-axis is inverted?

![](https://dl.dropboxusercontent.com/s/5j3dbeur5gawgz4/r84vsdev.jpg)

**Comment by darkwave** [comment]
I have the same issue with Blender 2.78.4 and latest version of the exporter.

Using following code, the scene appear mirrored. 

```
function init() {

  var loader = new THREE.ObjectLoader();
  loader.load("test.json", function(obj) {
    scene = obj;
    camera = obj.children[0];
    var ambient = new THREE.AmbientLight( 0xffffff );
  });
  
  renderer = new THREE.WebGLRenderer();
  renderer.setSize( window.innerWidth, window.innerHeight );

  document.body.appendChild( renderer.domElement );

}
```

![download](https://cloud.githubusercontent.com/assets/246174/23340800/b70cbbc8-fc3c-11e6-9eb3-5933682713a4.png)
![screenshot_2017-02-26_15-58-21](https://cloud.githubusercontent.com/assets/246174/23340801/b7107bdc-fc3c-11e6-9940-8b12c8f6e5a6.png)

**Comment by avallintine** [comment]
This seems related to https://github.com/mrdoob/three.js/issues/8511. The transformation matrix of objects is getting exported incorrectly in certain situations.

**Comment by daarong** [comment]
I just now downloaded the 86dev and I'm using its exporter with Blender 2.78c.  After importing into threejs 86dev as well as 84 master, I'm having to invert both the x and y parts of the vertices in order to correct the situation.

When I compare the .json file to one from a previous version, I can see clearly the X and Y components are inverted.

**Comment by bakajin** [comment]
I seem to have run into the same bug with the exporter from 86dev. Since I m exporting morphtargets. Which also required some working around. Leading up to this new problem.

So I would be keen to know how exactly you are inverting the vertices @daarong ?

**Comment by daarong** [comment]
@bakajin My 'fix' is just a hack too.  After my hack I am not seeing UV problems or inversion, or any symptoms really.

I'm using ObjectLoader.  Here is my TypeScript:

	mesh.updateMatrix();
	mesh.updateMatrixWorld(true);
	let geo = <THREE.Geometry>mesh.geometry;
	let verts = geo.vertices;
	for (let i = 0; i < verts.length; i++)
	{
		verts[i].setX(verts[i].x * -1);
		verts[i].setY(verts[i].y * -1);
	}
	geo.verticesNeedUpdate = true;

Good luck

**Comment by bakajin** [comment]
Well @daarong!

It seems to work a treat this hack. So massive thanks! Took me ages to figure out right. 

And it even seems to play nice with shape animations now. Which surprises me cause that is animating vertices. So I guess only the mesh vertices are effected ...

**Comment by Friksel** [comment]
Confirmed. I have the same issues here: x-axis are inverted. 
The error rises at the release-version of r85 (on blender 2.78c). 
On the previous exporter (r84) this problem wasn't not there, so it definitely got something to do with the r85 exporter.

**Comment by izar89** [comment]
Are you sure the issue still exists? The inverted x axis seems to be fixed. I still think something went wrong around r70 - r73 (the exporter in r71 and r72 throws errors). r70 gives me the result I would expect. But there could be a reason for this change I don't know of?

![](https://dl.dropboxusercontent.com/s/ratc5q36uv0cmuo/axis.jpg)

**Comment by Friksel** [comment]
@izar89 I just tested the r86 version you have inside the zip with my scene (which I load as a scene). It still messes up the scene completely, the same as r85. Older versions are like I described before. This was working in older versions. What I see now is a whole scene that is rotated/flipped and axis inside obects are flipped or moved too, but not always. Even the z-positions are sometimes wrong per object, but not on all objects the same.
Could this be something with a difference in objects where rotation and scale are applied in blender and other objects without having it applied before loading in the scene? Or maybe the exporter not respecting the objects origin points anymore?
Exporting your example blend-file is working together with your code, but there is obviously something different in my scene that wasn't a problem before, but is now in the latest exporters.

So to summarize: some objects are on the right position and have the right rotations, and some objects are completely off. So far I cannot find any structural logic to what's happening.

I'm using: "three": "^0.85.2", and "three-obj-loader": "^1.1.2" (latest now on npm).

**Comment by Friksel** [comment]
@izar89 I just did another test:

A
1) Exporting scene from blender through exporter-r84
2) Importing the exported json in the test you provided in the zip
3) Everything looks fine

B)
1) Exporting scene from blender through exporter-r86
2) Importing the exported json in the test you provided in the zip
3) Objects are messed up, like described in my previous post.

So we can eliminate the threejs version or my code, because I used your exact code now without any changes. Just a different json-export.

**Comment by Friksel** [comment]
btw, these are the settings I use:

http://imgur.com/a/vd2bp

**Comment by izar89** [comment]
You are right, it only gets flipped when you don't apply the scale and rotation. The issue is not resolved then. What do you think about the change between r70 and the next versions. For me r70 result is more logical.

![](https://dl.dropboxusercontent.com/s/otiav6zqwiaff4s/axisNoApply.jpg)

**Comment by manthrax** [comment]
There are a number of transformations and coordinate swaps that occur in the exporter, i think in attempts to address peoples preference for default "up" vector. These transformations are spread throughout the exporter, and don't seem completely consistent to me.

I have disabled them in a local branch of the exporter, and I get a solid 1 to 1 correspondence between blender and the three viewer.

You can get my modified exporter here if you want to check it out:
https://github.com/manthrax/three.js/tree/blender-exporter-remove-coordinate-mangling

**Comment by manthrax** [comment]
#11803

**Comment by Friksel** [comment]
@izar89 Great to see this point is picked up! Sorry for my late response. I wanted to provide you with a decent and helpful answer and wanted to look further into the exporter for that, but could't find the time and I'm afraid I'm not able very soon either. So for now my 2 cents:

I agree with @manthrax : it would be better to have a 1-to-1 export then to have an inconsistent export. Adding to this the locations of lights aren't exported correctly (or not at all?). I set these everytime by code after importing a scene from blender, and that feels silly and gives the exporter an unreliable feel. Especially because the behaviour seems to change after time when version changes.

Also other, to my opinion very importing things, like focus points of spotlights, are not exported, because these are marked out in the exporer-code. Can't really understand why, but could it be this all has got something to do with wanting to convert axis from blender to three too? So maybe this would also be solved when there would be a 1-to-1 export. 

Unfortunately I don't have time these days to help you with this, although I would like to, 'cause a better blender exporter is high on my list and I'm using it a lot, so sorry for this. But if you like my opinion:
having a reliable 1-to-1 export for all objects in the scene (including lights, cameras etc.) would be better then what the exporter is doing now, having to do all kind of corrections 'by hand' after import.
Keep up the good work! Again, like to see this is being picked up!

**Comment by manthrax** [comment]
@mrdoob:

I've gotten my head wrapped around the exporter pretty well, while adding support for multiple materials, on geometries and buffergeometries, and also adding multimaterial to the viewer.

For the exporter, I'd like to bring back the checkbox for YZ swap.. where when checked, you get the current broken behavior (unless I can fix it), and unchecked, you get pass through export.

I also want to fix the animation export to the point where both rest and pose export both work, without requiring a witch chant setup each time, where we can export multiple animated entities in one .blend file, and along with that I want to add animation preview support to the viewer.

Would this conflict with any other work going on, on the blender exporter or the three. viewer?

**Comment by manthrax** [comment]
I've also fixed a some other bugs in the exporter.. and a few bugs in the current viewer..
I updated the sematics of material.shading vs material.flatShaded in both the viewer and blender exporter..
If also fixed/updated the semantics of blending as well, as that seems to have been broken but unnoticed for a while.. in that the exporters are exporting values like "NormalBlending" but the three parser is expecting the shader constants themselves.

**Comment by WestLangley** [comment]
@manthrax Please do! :)

Here are my suggestions, FWIW. I have not looked at the exporter in years, so this is from memory...

1. Freeze the current exporter and start over with exporter 2.
2. Avoid the terminology "swapping YZ". The axes should not be swapped. The axes should just be rotated by a quarter-turn.
3. Remove all the code in the exporter that does any kind of coordinate swapping, and get the new exporter to work properly as a pass-through. @Mugen87 has developed a utility (#11540) that will rotate a scene. The 1/4-turn rotation can be applied in the loader callback as a temporary measure.
4. Finally, add back the coordinate toggle option by rotating the exported scene as a final step inside the loader -- instead of in the callback.
5. I would also advise avoiding the use of Euler angles -- specify the quaternion only.

**Comment by manthrax** [comment]
@WestLangley Agree 100%.

**Comment by mrdoob** [comment]
👍👍👍

**Comment by izar89** [comment]
@manthrax I created a very very basic blender exporter when I was testing the axis. Maybe it could be of any use for v2, you can find it here: https://github.com/izar89/Basic-ThreeJs-blender-exporter

**Comment by manthrax** [comment]
Thanks @izar89.. that is definitely helpful. Always good to get more specific examples of exporters, since it's  a little tricky to find good examples/docs ;). Appreciate it!

**Comment by pbonneville** [comment]
Just dropping a note that this flipped axis issue with the Blender export still seems to be an issue.

**Comment by Mugen87** [comment]
@pbonneville see #13130

**Comment by Mugen87** [comment]
The JSON Blender exporter has been removed with `R93` (#14117). 

Also see: https://threejs.org/docs/#manual/introduction/Loading-3D-models
